src/tools/manifest2proto.ts

```diff
@@ -35,12 +35,19 @@ export async function manifestToProtoPayload(manifest: Manifest) {
   return makeManifestProtoPayload(manifest.allParticles, manifest.allRecipes, manifest.allPolicies);
 }
 
-export async function encodePlansToProto(plans: Recipe[]) {
-  const specMap = new Map<string, ParticleSpec>();
-  for (const spec of flatMap(plans, r => r.particles).map(p => p.spec)) {
-    specMap.set(spec.name, spec);
-  }
-  return encodePayload(await makeManifestProtoPayload([...specMap.values()], plans, /* policies= */ []));
+export async function encodePlansToProto(plans: Recipe[], manifest: Manifest) {
+  // In the recipe data structure every particle in a recipe currently has its own copy
+  // of a particle spec. This copy is used in type inference and gets mutated as recipe
+  // is type checked. As we need to encode particle specs without such mutations, below
+  // we reach for ParticleSpecs from manifest.particles, instead of the ones hanging from
+  // the recipe.particles.
+  // This should be cleaned-up in the recipe data structure and type infrence code,
+  // once that happens, we can remove the below hack.
+  const specToId = (spec: ParticleSpec) => `${spec.name}:${spec.implFile}`;
+  const planParticleIds = flatMap(plans, p => p.particles).map(p => specToId(p.spec));
+  const particleSpecs = manifest.allParticles.filter(p => planParticleIds.includes(specToId(p)));
```

Comment:
What is the overlap between this round-about filtering and using `manifest.particles` instead of `.allParticles`?